### ISSUE TITLE
Asset Size In Bytes ProtoFlux Nodes

### ISSUE THREAD

**Issue by lxw404** [Yellow-Dog-Man/Resonite-Issues#1525]
### Is your feature request related to a problem? Please describe.

Currently it is difficult to impossible to estimate arbitrary static asset size.

### Describe the solution you'd like

There should be individual ProtoFlux nodes to return the current estimated size of any static asset given its current used variant.  This would allow for better understanding of how their size affects both performance and storage.

### Describe alternatives you've considered

The only asset that I am aware of that provides its estimated size in memory is the `StaticTexture2D` asset, however this is entirely within the inspector only which requires some unnecessary effort to extract.

### Requesters

LuxKitty

**Comment by lxw404** [comment]
Sorry, I am using "size" to refer to actual stored bytes, not the resolution.  The stored locations of assets vary along the lifecycle of the asset itself from its starting point on some network/storage device, to RAM, to VRAM as you mention, so I'll be a bit more explicit.

Specifically nodes for the following:
* Asset size when downloaded from its source - (This gives a general sense of network cost to the end user, and correlates strongly with it)
* Asset size when in RAM/VRAM at its current variant - (This gives a general sense of how much "capacity" an asset is taking up in their system resources and can give a general direction when pursuing improvements)

Use cases:
* Knowing the amount of bytes an asset takes to download can give a world designer a general sense of how much impact specific areas of a world will have on initial network bandwidth and generally correlates with how fast they load in.  This is beneficial for many purposes including deciding how to group assets to dynamically load them, or give the end user a better indication of which parts of the world are heavier on network download than others.
* Knowing VRAM/RAM usage is a clear use case for content creators.  If you know how much memory at a specific variant type that a texture/mesh takes up, you can get a better sense of the "budget" you have for your creation to fit within a target user's PC specs without forcing costly swaps between RAM/VRAM, and you can make better tradeoffs with explicitly choosing variants and resolutions to fit your needs.  For example, if you have an avatar that uses $N$ textures at resolutions, each at different compression levels, it is hard to know at a glance what is the most "costly" to VRAM budget.  (Of course this varies depending on an end user's setup, and their [settings](https://docs.google.com/document/d/1L9phRBabYLiihff_FYfzB8UoGaQRPFwfkG4S6HoqAYc/edit#heading=h.s4ll4dxguauj) which will affect this in the future, but it at least gives a good *upper ceiling* to shoot for rather than nothing and put the burden of thought on the people your creation affects).

Of course neither of these are ever going to give a full picture of "performance", but give at least an initial impression for end users to be more aware of in their own creations rather than being completely oblivious.  There are **many** other factors which complicate things, and these are only few of countless data points that could be exposed, but having more data to work with is always better than none.  Through strenuous testing, this data would already be extremely useful for creators to prioritize and collect empirical data points to improve their work.

**Comment by lxw404** [comment]
Both would be great.  ProtoFlux just allows users to build their own such profiling tools which can be reactive to such data.  For example, the grouping case I outlined above, having each asset enumerate its download size to a dynvar and then running a greedy algorithm on it in flux to partition slots into asset "groups" would be one such case for being able to access it directly.

**Comment by Frooxius** [comment]
For the size in VRAM/RAM of loaded asset, you can already get this information using the components in Assets/Utility, e.g. `Texture2DAssetMetadata`. I don't see reason to add them as nodes specifically, since you can just read out the values from this component easily.

For the download size, this is complicated for a few reasons:
- Usually the actual original asset isn't what gets downloaded. The asset variant system often splits the assets into multiple actual files and which ones and which type is downloaded can depend on a number of factors (which HW it's running on, settings and other requested variants). So using the size of the original asset could be very misleading from the actual download size
- Resonite uses deduplication too, so it can be a bit misleading on the download size - commonly used assets will not end up being downloaded in most of the cases, but they would show up in this
- The actual size of the asset isn't usually known at all in the engine even when the asset is loaded - this would require separate asynchronous fetch to the API to determine that

I think for the use cases you give it would be more beneficial to introduce more structured diagnostic and performance tools, e.g. to figure out how much users have to typically download for particular platforms.

**Comment by lxw404** [comment]
It's good to know that there is a component for texture asset metadata, however I don't know of any other components that expose other types of asset sizes in bytes, eg. Mesh, Document, Audio, etc.  I agree with the structured diagnostic and performance tools, however I would certainly side on enabling flexibility for users to build this into their own systems as well, whether component or ProtoFlux based.

As for download size, I do see there are levels of complexity to this which affect networking that could mislead.  Generally, the maximum impact is usually what someone coming into a completely new world will experience though, so this could serve as a decent upper ceiling baseline measurement that will only most likely be lower than measured.

**Comment by Frooxius** [comment]
We can potentially add more in-memory info for those. Not all of them have an easily calculable in-memory footprint though.

However the main point here is, neither components or ProtoFlux providing pieces of info on individual assets would work and provide you with flexibility - it would just give you misleading data - you are trying to break down complex contextual information into smaller pieces than it can actually be.

I'm very hesitant to introduce mechanisms that are more likely to be just misleading than not - sometimes very significantly so. E.g. you're assuming that the impact will be lower than measured, but this is not guaranteed - in a number of cases it can actually be higher. In some cases it can be very significantly lower. But most importantly - it's often irrelevant to the actual size of the download.

It's the wrong metric to try to optimize your worlds for. E.g. you could make it significantly lower by compressing your textures heavily with lossy compression vs using raw TGA/TIFF/PNG. But this will not actually affect the size of the download, because that's determined by the raw image data - you just lose quality for no reason.

This is one of the common misconceptions we've seen on Resonite and this would only perpetuate it.

**Comment by lxw404** [comment]
Agreed there are common misconceptions about optimization, however these are just one small missing piece of an entire very complex puzzle we currently have very opaque notions about at the moment.  I would rather transparency be ubiquitous and results measurable rather than it being a complete black box which perpetuates these misconceptions in the first place.  Especially people who try to equate other techniques to the way Resonite works.

We need abundantly more data and better ways of measuring results is the main point I am trying to make.

**Comment by Frooxius** [comment]
If you agree that they are common misconceptions, why ask for feature that will help perpetuate them and make people optimize for a metric that their content should absolutely not be optimized for?

You are stating that you'd want measurable results - but if you read my post in full, you'll see that this metric will not give you that.

At no point I suggested we keep things black box - rather I suggested that we implement features which would provide more comprehensive analysis and data instead, based on which creators could optimize their content and which will result in tangible benefits for performance. I'd be much happier to commit engineering time for that.

**Comment by lxw404** [comment]
That would be preferred.  I'm glad it will not be opaque, I was more interested to get any possible inclination of data rather than none.

**Comment by Dusty-Sprinkles** [comment]
@jae1911 this issue had information provided

**Comment by Dusty-Sprinkles** [comment]
@j4lc @DanteTucker This issue needs to be reopened
